File: packages/ivory-n5-client/helpers/ivory_n5.py
"""N5 HTTP helper for the existing execution contract; no local research state."""
import json
import os
import sys
import urllib.error
import urllib.parse
import urllib.request


class NoRedirect(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):
        return None


def request(path, body=None, key=None):
    headers = {"Accept": "application/json"}
    if body is not None:
        headers.update({"Content-Type": "application/json"})
        if key:
            headers.update({"Idempotency-Key": key})
    base = os.environ.get("IVORY_N5_SERVICE_URL", "http://127.0.0.1:4100").rstrip("/")
    req = urllib.request.Request(base + path, data=None if body is None else json.dumps(body).encode(), headers=headers)
    try:
        with urllib.request.build_opener(NoRedirect).open(req, timeout=30) as response:
            return json.load(response)
    except urllib.error.HTTPError as error:
        # Preserve service conflict payload; do not resolve conflicts locally.
        raise RuntimeError(json.dumps({"status": error.code, "body": json.load(error)})) from error


def submit(body, key):
    if not key or not key.strip():
        raise ValueError("Supply and retain an Idempotency-Key")
    return request("/v1/executions", body, key)


def get(execution_id):
    return request("/v1/executions/" + urllib.parse.quote(execution_id, safe=""))


def open(body):
    return request("/v1/projects/open", body)


def cite(body):
    return request("/v1/citations/resolve", body)


def run(body):
    return request("/v1/runspecs/resolve", body)


def edit(body, key):
    if not key or not key.strip():
        raise ValueError("Supply and retain an Idempotency-Key")
    return request("/v1/projects/edits", body, key)


if __name__ == "__main__":
    import builtins

    try:
        args = sys.argv[1:]
        if args == ["ready"]:
            result = request("/health/ready")
        elif len(args) == 2 and args[0] == "get":
            result = get(args[1])
        elif len(args) == 3 and args[0] == "submit":
            with builtins.open(args[1], encoding="utf-8") as source:
                result = submit(json.load(source), args[2])
        elif len(args) == 2 and args[0] in ("open", "cite", "run"):
            with builtins.open(args[1], encoding="utf-8") as source:
                result = {"open": open, "cite": cite, "run": run}[args[0]](json.load(source))
        elif len(args) == 3 and args[0] == "edit":
            with builtins.open(args[1], encoding="utf-8") as source:
                result = edit(json.load(source), args[2])
        else:
            raise ValueError("Usage: ivory_n5.py ready | get ID | submit REQUEST.json KEY | open|cite|run REQUEST.json | edit REQUEST.json KEY")
        print(json.dumps(result))
    except Exception as error:
        print(str(error), file=sys.stderr)
        sys.exit(1)

File: packages/ivory-n5-client/helpers/ivory_n5.R
# N5 HTTP helper. JSON stays text to preserve arrays/nulls/numeric precision.
# Requires httr2; never starts a service or writes records.
ivory_request <- function(path, body = NULL, key = NULL) {
    if (!requireNamespace("httr2", quietly = TRUE)) stop("Install httr2 in the candidate R environment")
    base <- sub("/$", "", Sys.getenv("IVORY_N5_SERVICE_URL", "http://127.0.0.1:4100"))
    req <- httr2::request(paste0(base, path))
    req <- httr2::req_headers(req, Accept = "application/json")
    req <- httr2::req_options(req, followlocation = FALSE)
    req <- httr2::req_timeout(req, 30)
    req <- httr2::req_error(req, is_error = function(resp) FALSE)
    if (!is.null(body)) {
        if (is.null(key) || !nzchar(trimws(key))) stop("Supply and retain an Idempotency-Key")
        req <- httr2::req_headers(req, `Idempotency-Key` = key)
        if (!is.character(body) || length(body) != 1L) stop("Pass the complete request as JSON text")
        req <- httr2::req_body_raw(req, body, type = "application/json")
    }
    resp <- httr2::req_perform(req)
    payload <- httr2::resp_body_string(resp)
    if (httr2::resp_status(resp) >= 300) {
        stop(structure(list(message = paste("Service HTTP", httr2::resp_status(resp)),
                            status = httr2::resp_status(resp), body = payload),
                       class = c("ivory_http_error", "error", "condition")))
    }
    payload
}
ivory_submit <- function(body, key) ivory_request("/v1/executions", body, key)
ivory_get <- function(id) ivory_request(paste0("/v1/executions/", utils::URLencode(id, reserved = TRUE)))

if (sys.nframe() == 0L) {
    args <- commandArgs(trailingOnly = TRUE)
    result <- if (identical(args, "ready")) ivory_request("/health/ready") else if (length(args) == 2L && args[1] == "get") {
        ivory_get(args[2])
    } else if (length(args) == 3L && args[1] == "submit") {
        ivory_submit(paste(readLines(args[2], warn = FALSE, encoding = "UTF-8"), collapse = "\n"), args[3])
    } else stop("Usage: ivory_n5.R ready | get ID | submit REQUEST.json KEY; N5 project/citation/RunSpec/edit contracts are blocked")
    cat(result, "\n")
}
